feat: add date adapter for luxon

crisbeto opened the pull request:
Adds the `LuxonDateAdapter` that can be used to work with Luxon dates in the `MatDatepicker`.

mmalerba commented on src/material-luxon-adapter/BUILD.bazel:
```diff
@@ -0,0 +1,26 @@
+
+load("//tools:defaults.bzl", "ng_module", "ng_package")
+load("//:packages.bzl", "ROLLUP_GLOBALS")
+
```
Can probably update this now, based on https://github.com/angular/material2/pull/14687

mmalerba commented on src/material-luxon-adapter/adapter/luxon-date-adapter.ts:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  getFirstDayOfWeek(): number {
+    return 0;
```
Maybe leave a comment about Luxon not supporting this

mmalerba commented on src/material-luxon-adapter/adapter/luxon-date-adapter.ts:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  clone(date: DateTime): DateTime {
+    return date;
```
doesn't look like this is actually cloning?

crisbeto replied:
We don't need to clone since the Luxon `DateTime` is immutable.

mmalerba replied:
I think we still ought to, otherwise `d === adapter.clone(d)` which feels wrong

crisbeto commented:
I've addressed the feedback @mmalerba. Also @devversion can you take a look at the Bazel setup? I've based it off of what you did for the Moment adapter.

devversion commented on src/material-luxon-adapter/BUILD.bazel:
```diff
@@ -0,0 +1,56 @@
+  deps = [
+    "@angular//packages/core",
+    "@matdeps//@types/luxon",
+    "@matdeps//luxon",
```
We don't need this here (only the types). Moment ships its typings as part of `@matdeps//moment`

devversion commented on src/material-luxon-adapter/BUILD.bazel:
```diff
@@ -0,0 +1,56 @@
+  # Explicitly specify the tsconfig that is also used by Gulp. We need to explicitly use this
+  # tsconfig because in order to import Luxon with TypeScript, we need some special options
+  # enabled.
+  tsconfig = ":tsconfig-build.json",
```
Maybe my commenting wasn't too clear, but we shouldn't need this here because Luxon doesn't need the `synthetic` imports mentioned in the `tsconfig-build.json` of the moment adapter package.

devversion commented on src/material-luxon-adapter/BUILD.bazel:
```diff
@@ -0,0 +1,56 @@
+    "//src/lib/testing",
+    "//src/cdk/platform",
+    "@matdeps//@types/luxon",
+    "@matdeps//luxon",
```
Same as above. Should be able to remove this from here.

crisbeto commented:
@jelbourn @mmalerba I've addressed the feedback. Can you take another look?

mmalerba commented:
**Caretaker Note:** `material-luxon-adapter` should not be synced into g3, may needto update copybara

mmalerba commented:
Fixes #12321, #10427

andrewseguin commented:
Blocked pending further discussion on our strategy for supporting multiple date adapters

DaSchTour commented:
Hope this discussion won't take too long 😕

jturekburnett commented on src/material-luxon-adapter/adapter/luxon-date-adapter.ts:
```diff
@@ -0,0 +1,246 @@
+  }
+
+  getDayOfWeekNames(style: 'long' | 'short' | 'narrow'): string[] {
+    return Info.weekdays(style, {locale: this.locale});
```
Not sure if should be changed here or somewhere else, but the way this is, the weekdays on the month view of the calendar don't line up correctly with the dates because it starts on Monday instead of Sunday.  For example, April 1, 2019 was on a Monday, but this adapter causes the 'T' for Tuesday to appear above April 1 instead of the 'M'.

dmitry-urenev commented on src/material-luxon-adapter/adapter/luxon-date-adapter.ts:
```diff
@@ -0,0 +1,246 @@
+    return Info.months(style, {locale: this.locale});
+  }
+
+  getDateNames(): string[] {
```
this function might be implemented like this:

```suggestion
 public getDateNames(): string[] {
        return range(31, day =>
            new DateTime()
                .set({ month: 1, day })
                .setLocale(this.locale)
                .toFormat('d')
        );
}
```

LosD commented:
> Blocked pending further discussion on our strategy for supporting multiple date adapters
@andrewseguin Did you ever complete that discussion?

dmitrye commented:
It would be good for someone to finish this adapter. The current method I had to resort to was to make properties for each Date Input

The html looks like this:
```
<input matInput id="endDate" formControlName="endDate" name="endDate" placeholder="Select a Date" autocomplete="off"
                               [value]="endDateAsDate" type="text" class="form-control" [matDatepicker]="myEndDatePicker" (focus)="myEndDatePicker.open()">
```
The .ts file looks like this (converts Luxon DateTime to JSDate/Date object):
```
  get endDateAsDate(){
    return <Date>(<DateTime>this.f.endDate.value).toJSDate();
  }
```

It's only needed to set the default value coming back from DB or if you want to preset it to Today. But remembering to do this everywhere doesn't seem sustainable.

Enngage commented:
+ Luxon date adapter

weihsth commented:
+ Luxon date adapter

DaSchTour commented:
I think with google maps and youtube player as new modules in this repository there should not be any doubt about adding this module here.

sbarbat commented:
Hi guys, what is needed to merge this? Happy to help here 😄

LosD commented:
> Hi guys, what is needed to merge this? Happy to help here 😄

@sbarbat unfortunately, it is becoming a standard pattern for this project: Things moving forward nicely, then @andrewseguin coming in with some "we need to discuss this", that discussion apparently never taking place, and everything dying from there on out.

I fully understand the need to design things properly, but if you decide on a discussion, effing HOLD that discussion instead of just letting stuff die for years.

jelbourn commented:
Status us tracked here: https://github.com/angular/components/issues/20599#issuecomment-705201800

The gist is that there's debate on the team as to whether we should support these adapters, or defer to community implementations.

mtpultz commented:
@andrewseguin @jelbourn would it not make sense to support at least one new date adapter while you internally decide whether you want to maintain date adapters in the Angular repo, or push these adapters out for the Angular community to maintain?  By allowing this PR to continue and be merged in you're not locking Angular into maintaining date adapters.

Supporting Luxon in Angular makes sense since you currently support a Moment date adapter so all you are doing is providing an update due to a deprecation, which is good practice and a maintenance task.  Luxon is contained within the Moment repo making it essentially Moment v2 so arguably you can still have the discussion and split out the Luxon date adapter into a community repo afterwards, but by having this on hold and opening an internal discussion on whether Angular should or should not support date adapters (that apparently hasn't happened) doesn't stop new projects from perpetuating the use of Moment.

@andrewseguin https://github.com/angular/components/pull/14681#issuecomment-457685311 you blocked a PR that was apparently marked ready to merge in January 2020, and had been open since December 2018.  With the offers to help from the community and the amount of work already put into this PR this could have been completed and pushed in leaving all sorts of time for an internal discussion that the community isn't privy on where date adapters should live, and it would solve the issue we're having with a deprecated dependency being the only officially supported fluent date adapter in Angular.

@jelbourn https://github.com/angular/components/issues/20599#issuecomment-705201800 is an opinion PR that discusses which date library should have a date adapter, but you can't discuss what libraries should have date adapters if internally no decision has been made on whether they should be maintained by Angular or the community.

This PR has support from the community.  Please take the hold off to remove the bottleneck on this date adapter so we can adopt a new date adapter, drop a deprecated dependency, and then where everything ends up living can be decided.

andreialecu commented:
I ended up needing this date adapter as well, and just copied the code from this PR and published it on npm: https://www.npmjs.com/package/ngx-material-luxon

Seems to work for my purposes.

jelbourn commented:
I would personally like to move forward with this, but I don't have unilateral decision-making power when it comes to publishing new `@angular` packages, and unfortunately there isn't consensus from the rest of the Angular team that this is something we should work on yet. Overall the team will be in a better position to consider additions like this once ViewEngine (the pre-Ivy rendering engine) is deleted.

Polyterative commented:
any progress on this?

wiegell commented:
Any news?

crisbeto commented:
Closing in favor of #23167.
